GUACAMOLE-465: beginning to support new codecs and containers

sreid8 opened the pull request:
This patch adds a dependency to "libavformat", the ffmpeg library that handles writing video and audio streams properly into video/audio containers. The patch makes NO changes to the way that guacenc functions from the user perspective. 

This change is required to eventually support more codecs and containers. Previously, guacenc wrote encoded video frames directly into an output file. This works perfectly file for mpeg4 in an m4v container, but will not work for most other combinations (like h.264 in mp4, for example). However, this patch does not add any support for any new codecs or containers. After this patch, guacenc will still _only_ support mpeg4 in m4v. 

This is patch 1 of 3 that will be provided for GUACAMOLE-465. The code for the other patches is written already on a branch that I've been maintaining, but in the interest in getting the functionality merged, I'm breaking it into multiple patches to make it easier to review.
**The current plans for the next patches are as follows:**
- Patch 2 will add the ability for the user to decide to specify input and output files rather that using the current batch interface, but will still provide the batch interface.
- Patch 3 will add h.264, VP8, and h.265 codecs in addition to the already existing mpeg4 and will also add mp4 and webm container support in addition to the already existing m4v support.

I welcome any suggestions for changes for this to be merged and will answer any questions about decisions I made while writing the patch.  This _will_ require changes to the Guacamole Manual, but I didn't see those documents in the repo, so I haven't modified them myself. If those documents are open and accept pull requests, I will gladly update them myself as part of this patch.

mike-jumper commented on src/guacenc/ffmpeg-compat.c:
```diff
@@ -165,3 +197,62 @@ int guacenc_avcodec_encode_video(guacenc_video* video, AVFrame* frame) {
+        int pix_fmt,
+        AVRational time_base) {
+#if LIBAVFORMAT_VERSION_INT < AV_VERSION_INT(57, 33, 100)
+    stream->codec->bit_rate = bitrate;
```
Why is the `codec` parameter ignored in this case?

sreid8 replied:
In video.c, the stream is allocated already with the exact same codec that is passed here. It was only after libavformat 57.33.100 where the codec output parameters of the video were promoted from ```AVStream->codec->*``` to ```AVStream->codecpar```. The new function ```guacenc_open_avcodec``` handles moving the parameters from the ```context``` (which are set here) to ```codecpar``` in the case that libavformat >= 57.33.100. This ffmpeg API change is documented [here](https://github.com/FFmpeg/FFmpeg/blob/master/doc/APIchanges#L481).
However, before 57.33.100, the ```stream->codec``` is where the parameters are placed and the ```codec``` member of this struct has already been set to the proper codec when it was allocated.

mike-jumper commented on src/guacenc/ffmpeg-compat.c:
```diff
@@ -165,3 +197,62 @@ int guacenc_avcodec_encode_video(guacenc_video* video, AVFrame* frame) {
+        context->qmin = qmin;
+        context->pix_fmt = pix_fmt;
+        context->time_base = time_base;
+        stream->time_base = time_base;
```
Can you describe why this function is touching `stream` as a side effect?

sreid8 replied:
Similar to my other comment above about ```codec``` being ignored, it seems like ffmpeg is doing some soul searching about the roles of ```AVCodecContext``` and ```AVStream```. Just like the way that codec parameters are in ```AVStream->codecpar``` in >=57.33.100 instead of ```AVCodecContext->(parameter)``` or ```AVStream->codec->(parameter)```, it seems like they're having a hard time deciding where the ```time_base``` should be set. So since there are some versions where it needs to be on the ```AVCodecContext``` and some on the ```AVStream```, but no cases where it _can't_ be on both, I decided to put it on both to cover all our bases.

mike-jumper replied:
Alrighty. I think it's worth noting in the documentation of `guacenc_build_avcodeccontext()` that the `AVStream` will be modified (and the nature of that modification), but I'm OK with it being done.

mike-jumper commented on src/guacenc/ffmpeg-compat.c:
```diff
@@ -51,8 +51,41 @@
+    pkt->size = size;
+    pkt->stream_index = video->output_stream->index;
+    ret = av_interleaved_write_frame(video->container_format_context, pkt);
+    free(pkt);
```
Why dynamically allocate an `AVPacket` only to immediately free it after use?

sreid8 replied:
Great point. Changed to a stack variable.

mike-jumper commented on src/guacenc/ffmpeg-compat.c:
```diff
@@ -165,3 +197,62 @@ int guacenc_avcodec_encode_video(guacenc_video* video, AVFrame* frame) {
 }
 
+AVCodecContext* guacenc_build_avcodeccontext(AVStream* stream,
+        AVCodec* codec,
```
Please follow the general style of the surrounding code. This function stands out as the only one to dedicate a separate line to each parameter, regardless of line length.

sreid8 replied:
My mistake! I write Java all day at work and so the only excuse I can think of for why I didn't see this when I was eyeball-linting is that I must be too used to how long Java parameters tend to be pretty long. Consider it fixed. :disappointed:

sreid8 commented:
I added a commit to fix some of the issues you brought up, Mike, and I also commented on some issues as well. If you have any further concerns after the comments I wrote, let me know

sreid8 commented:
Hi @mike-jumper! It's been a few months since there was any activity on this, but I'm still interested in contributing it and the other patches I've got lined up. I know it's probably a low priority for the guacamole project as a whole, but is there anything I can do to help get this functionality merged?

mike-jumper commented:
Sorry about that - I'll try to continue the review.

h0tw1r3 commented:
@mike-jumper will this be merged at some point?

mgd134 commented:
Does this PR add the ability to have audio in guacamole RDP session recordings? My RDP sessions have audio while I'm using them, but after running `guacenc` to make a video of my session, the resulting video does not have audio. Or am I perhaps doing something incorrectly?

necouchman commented:
> Does this PR add the ability to have audio in guacamole RDP session recordings? My RDP sessions have audio while I'm using them, but after running guacenc to make a video of my session, the resulting video does not have audio. Or am I perhaps doing something incorrectly?

I don't think so - I think this PR deals primarily with adding codecs to `guacenc` to be able to output the videos in other formats.  For audio support in `guacenc`, please start by asking your question on the mailing list - if it's determined that this needs to be added as a feature you can enter a JIRA issue for it.

http://guacamole.apache.org/support/

sreid8 commented:
> Does this PR add the ability to have audio in guacamole RDP session recordings? My RDP sessions have audio while I'm using them, but after running `guacenc` to make a video of my session, the resulting video does not have audio. Or am I perhaps doing something incorrectly?

@necouchman is right, this PR doesn't have audio support, no. It's something I'm interested in adding for my own use at some point and I'd be likely to offer it back.

mike-jumper commented on src/guacenc/ffmpeg-compat.c:
```diff
@@ -165,3 +194,56 @@ int guacenc_avcodec_encode_video(guacenc_video* video, AVFrame* frame) {
+    stream->codec->pix_fmt = pix_fmt;
+    stream->codec->time_base = time_base;
+#if LIBAVFORMAT_VERSION_INT < AV_VERSION_INT(55, 44, 100)
+    stream->codec->time_base = time_base;
```
This particular line seems to be here twice.

sreid8 replied:
After looking at it, it seems like the `#else` is the only really important part, so I re-did the boolean so that there was no duplicated line.

mike-jumper commented on src/guacenc/ffmpeg-compat.c:
```diff
@@ -165,3 +194,56 @@ int guacenc_avcodec_encode_video(guacenc_video* video, AVFrame* frame) {
+        AVCodec *codec, AVDictionary **options,
+        AVStream* stream) {
+    int ret = 0;
+    ret = avcodec_open2(avcodec_context, codec, options);
```
The initialized value for `ret` is discarded here. Why not `int ret = avcodec_open2(...)`?

mike-jumper commented on src/guacenc/ffmpeg-compat.c:
```diff
@@ -165,3 +194,56 @@ int guacenc_avcodec_encode_video(guacenc_video* video, AVFrame* frame) {
+    int codecpar_ret = 0;
+    /* copy stream parameters to the muxer */
+    codecpar_ret = avcodec_parameters_from_context(stream->codecpar,
+            avcodec_context);
```
Same here - as `codecpar_ret` is initialized, only to have that value immediately overwritten by the return value of `avcodec_parameters_from_context()`, it would be better to just initialized the variable with that return value directly.

mike-jumper commented on src/guacenc/ffmpeg-compat.h:
```diff
@@ -78,5 +78,70 @@
+ * of libavformat installed
+ *
+ * @param stream
+ *     The open AVStream
```
The format Doxygen-style comments that are used within guacamole-server to document the various functions is similar to JavaDoc and JSDoc in the need for periods, etc. The descriptions of parameters and the description of the function all need to end with a final `.`. For example:

```c
/**
 * Creates and sets up the AVCodecContext for the appropriate version
 * of libavformat installed.
 *
 * @param stream
 *     The open AVStream.
```

mike-jumper commented on src/guacenc/ffmpeg-compat.h:
```diff
@@ -78,5 +78,70 @@
+ */
+AVCodecContext* guacenc_build_avcodeccontext(AVStream* stream, AVCodec* codec,
+        int bitrate, int width, int height, int gop_size, int qmax, int qmin,
+        int pix_fmt, AVRational time_base);
```
Looks like the documentation for the return value is missing here.

mike-jumper commented on src/guacenc/ffmpeg-compat.h:
```diff
@@ -78,5 +78,70 @@
+ *               lower versions
+ *
+ * @return zero on success, a negative value on error
+ */
```
Please use the same comment style as the other recent functions within guacamole-server (the style you aimed at already with `guacenc_build_avcodeccontext()`). The descriptions of the accepted parameters should start on the next line following the `@param` line, indented by one level, with a blank line between each parameter description.

mike-jumper commented on src/guacenc/video.h:
```diff
@@ -42,16 +50,23 @@
     /**
-     * Output file stream.
+     * AVStream for video output.
+     * Persists via the AVFormatContext.
```
I'm not sure what you mean here by "Persists via the AVFormatContext." Can this comment be clarified?

sreid8 replied:
Reworded to be more accurate and clear.

mike-jumper commented on src/guacenc/ffmpeg-compat.c:
```diff
@@ -51,8 +51,38 @@
+
+#if LIBAVCODEC_VERSION_INT < AV_VERSION_INT(54,1,0)
+    AVPacket pkt;
+    /* have to create a packet around the encoded data we have */
```
This isn't a deal-breaker for me, but as a point of consistency in style, this comment (`/* all lowercase */`) doesn't match the style of existing comments (`/* Sentence case */`). I would recommend making things consistent.

mike-jumper commented on src/guacenc/video.c:
```diff
@@ -51,25 +69,35 @@ guacenc_video* guacenc_video_alloc(const char* path, const char* codec_name,
-    context->gop_size = 10;
-    context->max_b_frames = 1;
-    context->pix_fmt = AV_PIX_FMT_YUV420P;
+    //if format needs global headers, write them
```
Single-line comments within guacamole-server should be C-style (`/* ... */`) rather than C++-style (`// ...`).

sreid8 commented:
Hi @mike-jumper. I think I've addressed the comments you had here. I do think it's important that I call out a few changes that I made outside of the requested changes, however. Over the time that this PR has been up, some things had to change, so I updated them.

1. I wrapped the initialization of libavformat in a [similar guard](https://github.com/sreid8/guacamole-server/blob/guacenc_libavformat/src/guacenc/guacenc.c#L84-L86) to the way that the initialization was wrapped for libavcodec because the initialization calls have been deprecated in newer versions of FFmpeg.

2. I added a new macro to ffmpeg-compat.h to handle the change libavcodec made to the `CODEC_FLAG`s. Previously, these were all `CODEC_FLAG_*` format, but now they're `AV_CODEC_FLAG_*` format, so the new macro [`GUACENC_FLAG_GLOBAL_HEADER`](https://github.com/sreid8/guacamole-server/blob/guacenc_libavformat/src/guacenc/ffmpeg-compat.h#L59-L63) resolves that issue.

necouchman commented:
From a completely logistical point of view, you've merged the master code into this branch, which has resulted in an extra commit in your pull request (f058286).  My guess is that you were trying to re-incorporate the upstream changes, which is good, but the way to do this is to rebase your branch on top of master, and handle any conflicts.  So, you'd:
* Make sure your fork of the master branch is up-to-date ("git pull --ff-only upstream/master", where upstream is either the upstream repository or the github.com/apache/guacamole-server mirror)
* Checkout your branch (guacenc_libavformat)
* git rebase master
* Once you've handled all of the conflicts and such, do a force push of that branch.

Let me know if you need any help with that - it can be a bit challenging the first time or two.

sreid8 commented:
@necouchman yeah, I was trying to make sure I could test with everything that was on master. I'll go back through and rebase to correct it

mike-jumper commented on src/guacenc/ffmpeg-compat.h:
```diff
@@ -78,5 +88,81 @@
+ * the AVStream will also be affected by having its time_base field
+ * set to the value passed into this function.
+ *
+ * @param stream
+ *     The open AVStream
```
As with JavaDoc, the formatting of these Doxygen comments requires that the description of parameters, return values, etc. all end with periods. This is correct for the other new code - some sneaky creature just went through this one and ate all the periods. ;)

necouchman replied:
Ping @sreid8 - any chance you could finish this up?  Looks like it's about ready to merge so we can close out for next release.

mike-jumper commented:
@sreid8 - I went ahead and pulled your commits into a new PR (#286) and quickly made the changes to address my own nitpicks. Thanks, and sorry again that it's taken me so long to do my part to move this forward.

sreid8 commented:
@mike-jumper no problem at at all. Glad to see #286 got merged, and I took a look at the commit that was added there so that the next PR is closer to ready from the start. 

Thanks for accepting it. I'm excited to keep contributing!

sjxz12358 commented:
> > 此 PR 是否添加了在 guacamole RDP 会话录音中添加音频的功能？我的 RDP 会话在我使用它们时有音频，但是在运行`guacenc`我的会话视频后，生成的视频没有音频。或者我可能做错了什么？
> 
> @necouchman是的，这个 PR 没有音频支持，没有。这是我有兴趣在某个时候添加以供我自己使用的东西，我很可能会提供回来。

 still not support it?
